rdtools/soiling.py

```diff
@@ -1052,3 +1065,1552 @@ def monthly_soiling_rates(soiling_interval_summary, min_interval_length=14,
+        order = ('SR', 'SC' if degradation_method == 'STL' else 'Rd')
+        t0 = time.time()
+        bt_kdfs, bt_SL, bt_deg, knobs, adfs, RMSEs, SR_is_1, rss, errors = \
+            [], [], [], [], [], [], [], [], ['Bootstrapping errors']
+        for b in range(reps):
+            try:
+                # randomly choose model knobs
+                dt = np.random.uniform(knob_alternatives[1][0],
+                                       knob_alternatives[1][-1])
+                pt = np.random.uniform(knob_alternatives[2][0],
+                                       knob_alternatives[2][-1])
+                pn = np.random.uniform(process_noise / 1.5, process_noise * 1.5)
+                renormalize_SR = np.random.choice([None,
+                                                   np.random.uniform(.5, .95)])
+                ffill = np.random.choice([True, False])
+                knobs.append([dt, pt, pn, renormalize_SR, ffill])
+
+                # Sample to infer soiling from
+                bootstrap_sample = \
+                    all_bootstrap_samples[b] / seasonal_samples[b]
+
+                # Set up a temprary instance of the CODSAnalysis object
+                temporary_cods_instance = CODSAnalysis(bootstrap_sample)
+
+                # Do Signal decomposition for soiling and degradation component
+                kdf, results_dict = temporary_cods_instance.iterative_signal_decomposition(
+                        max_iterations=4, order=order, clip_soiling=True,
+                        detection_tuner=dt, pruning_iterations=1,
+                        pruning_tuner=pt, process_noise=pn,
+                        renormalize_SR=renormalize_SR, ffill=ffill,
+                        degradation_method=degradation_method, **kwargs)
+
+                # If we can reject the null-hypothesis that there is a unit
+                # root in the residuals:
+                if results_dict['adf_res'][1] < .05:  # Save the results
+                    bt_kdfs.append(kdf)
+                    adfs.append(results_dict['adf_res'][0])
+                    RMSEs.append(results_dict['RMSE'])
+                    bt_deg.append(results_dict['degradation'])
+                    bt_SL.append(results_dict['soiling_loss'])
+                    rss.append(results_dict['residual_shift'])
+                    SR_is_1.append((kdf.soiling_ratio == 1).mean())
+                else:
+                    seasonal_samples.drop(columns=[b], inplace=True)
+
+            except ValueError as ve:
+                seasonal_samples.drop(columns=[b], inplace=True)
+                errors.append([b, ve])
+
+            # Print progress
+            if verbose:
+                _progressBarWithETA(b+1, reps, time.time()-t0, bar_length=30)
+
+        # Reweight and save weights
+        weights = 1 / np.array(RMSEs) / (1 + np.array(SR_is_1))
+        weights /= np.sum(weights)
+        self.knobs_n_weights = pd.concat(
```

Comment:
note this brings up `FutureWarning: In a future version of pandas all arguments of concat except for the argument 'objs' will be keyword-only
  concat_tot_mod = pd.concat([kdf.total_model for kdf in bt_kdfs], 1)`

Reply:
Can wait for a follow up PR